desc/objectives/_generic.py

```diff
@@ -15,6 +19,262 @@
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization.
+    name : str, optional
+        Name of the objective function.
+
+    # TODO: add example
+
+    """
+
+    _units = "(Unknown)"
+    _print_value_fmt = "External objective value: {:10.3e}"
+
+    def __init__(
+        self,
+        eq,
+        fun,
+        dim_f,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=False,
+        normalize_target=False,
+        loss_function=None,
+        fd_step=1e-4,  # TODO: generalize this to allow a vector of different scales
+        vectorized=False,  # False or int
+        name="external",
+        **kwargs,
+    ):
+        assert isinstance(vectorized, bool) or isinstance(vectorized, int)
+        if target is None and bounds is None:
+            target = 0
+        self._eq = eq.copy()
+        self._fun = fun
+        self._dim_f = dim_f
+        self._fd_step = fd_step
+        self._vectorized = vectorized
+        self._kwargs = kwargs
+        if self._vectorized:
+            try:  # spawn a new environment so the backend can be set to numpy
+                mp.set_start_method("spawn")
+            except RuntimeError:  # context can only be set once
+                pass
+        super().__init__(
+            things=eq,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode="fwd",
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        self._scalar = self._dim_f == 1
+        self._constants = {"quad_weights": 1.0}
+
+        if self._vectorized:
+            max_processes = (
+                self._vectorized
+                if isinstance(self._vectorized, int)
+                else os.cpu_count()
+            )
+            self._pool = mp.Pool(processes=max_processes)
+
+        def fun_wrapped(params):
+            """Wrap external function with possibly vectorized params."""
+            # number of equilibria for vectorized computations
+            param_shape = params["Psi"].shape
+            num_eq = param_shape[0] if len(param_shape) > 1 else 1
+
+            if self._vectorized and num_eq > 1:
+                # convert params to list of equilibria
+                eqs = [self._eq.copy() for _ in range(num_eq)]
+                for k, eq in enumerate(eqs):
+                    # update equilibria with new params
+                    for param_key in self._eq.optimizable_params:
```

Comment:
instead of this I think you can use `tree_unstack` from backend and to turn a dict of arrays into a list of dict, and then just do `eq.params_dict = params[i]`

Reply:
Normally `params` is a dict of 1D arrays. When vectorized, it becomes a dict of 2D arrays, where each row corresponds to the param values for a different equilibrium. So I don't think the pytree utils will help here.